pkg/operator/test/volumes.go

```diff
@@ -0,0 +1,227 @@
+	} else if v.VolumeSource.HostPath != nil {
+		sourceString = v.VolumeSource.HostPath.Path
+	} else {
+		// If can't convert the vol source to something human readable, just output something useful
+		sourceString = fmt.Sprintf("%v", v.VolumeSource)
+	}
+	return fmt.Sprintf("{%s : %s}", v.Name, sourceString)
+}
+
+// HumanReadableVolumeMounts returns a string representation of a list of Kubernetes volume mounts which
+// is more compact and human readable than the default string go prints.
+func HumanReadableVolumeMounts(mounts []v1.VolumeMount) string {
+	stringMounts := []string{}
+	for _, mount := range mounts {
+		stringMounts = append(stringMounts, HumanReadableVolumeMount(&mount))
+	}
+	return fmt.Sprintf("%v", stringMounts)
+}
+
+// HumanReadableVolumeMount returns a string representation of a Kubernetes volume mount which is more
+// compact and human readable than the default string go prints.
+func HumanReadableVolumeMount(m *v1.VolumeMount) string {
+	return fmt.Sprintf("{%s : %s}", m.Name, m.MountPath)
+}
+
+/*
+ * Fully-implemented test for matching Volumes and VolumeMounts
+ */
+
+// VolumesSpec is a struct which includes a list of Kubernetes volumes as well as additional
+// metadata about the volume list for better identification during tests.
+type VolumesSpec struct {
+	// Moniker is a name given to the list to help identify it
+	Moniker string
+	Volumes []v1.Volume
+}
+
+// MountsSpec is a struct which includes a list of Kubernetes volume mounts as well as additional
+// metadata about the volume mount list for better identification during tests.
+type MountsSpec struct {
+	// Moniker is a name given to the list to help identify it
+	Moniker string
+	Mounts  []v1.VolumeMount
+}
+
+// VolumesAndMountsTestDefinition defines which volumes and mounts to test and what those values
+// should be. The test is intended to be defined with VolumesSpec defined as a pod's volumes, and
+// the list of MountsSpec items defined as the volume mounts from every container in the pod.
+type VolumesAndMountsTestDefinition struct {
+	VolumesSpec     *VolumesSpec
+	MountsSpecItems []*MountsSpec
+}
+
+// TestMountsMatchVolumes tests two things:
+// (1) That each volume mount in each every MountsSpec has a corresponding volume to source it
+//     in the VolumesSpec
+// (2) That there are no extraneous volumes defined in the VolumesSpec that do not have a
+//     corresponding volume mount in any of the MountsSpec items
+func (d *VolumesAndMountsTestDefinition) TestMountsMatchVolumes(t *testing.T) {
+	// Run a test for each MountsSpec item to verify that all the volume mounts within each item
+	// have a corresponding volume in VolumesSpec to source it
+	for _, mountsSpec := range d.MountsSpecItems {
+		t.Run(
+			fmt.Sprintf("%s mounts match volumes in %s", mountsSpec.Moniker, d.VolumesSpec.Moniker),
+			func(t *testing.T) {
+				for _, mount := range mountsSpec.Mounts {
+					if err := VolumeExists(mount.Name, d.VolumesSpec.Volumes); err != nil {
```

Comment:
you could use `assert.Nil(t, err)`

Reply:
It seems to me that there is a need for more explicit error messages here because when the line number of the error is printed, it is the line from this file, which doesn't give much indication of what exactly it is looking at that is wrong at in the file which calls this `.../mgr/spec_test.go` or `.../mon/spec_test.go`, for example.

Reply:
Oh wait, you said above that I can do `assert.Nil(t, err, "error msg %v", arg ...)`. I could for sure do this, yes.

Reply:
when the error is not nil, it will actually print what the error message is, so I think you'll find you already get the info you want. Similarly, other types of asserts will usually have a useful message by default. Not always, but frequently.